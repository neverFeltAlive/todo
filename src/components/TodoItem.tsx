import React, {useState} from 'react';
import {TodoItemType} from "./TodoList";
import {FaTimes, FaCheck} from "react-icons/fa";
import styled from "styled-components";
import {Button as Btn} from "./UI";

//region Type
type TodoItemProps = {
    item: TodoItemType
    removeFunc: (id: number) => void
    toggleComplete: () => void
}

type StyledProps = {
    isComplete: boolean
}
//endregion

//region Style
const Button = styled(Btn)`
  color: lightcoral;
  cursor: pointer;
  
  @media (min-width: 1200px){
    margin-left: 20px;
  }
`;

const Container = styled.div<StyledProps>`
  display: flex;
  align-items: center;
  justify-content: space-between;
  color: whitesmoke;
  font-weight: 700;
  padding: 0 5px;
  font-size: 18px;
  border: 3px solid transparent;
  cursor: pointer;

  @media (min-width: 1200px){
    font-size: 25px;
  }
  
  ${props => {
    if (props?.isComplete){
      return `
            opacity: 0.5;
        `
    } else return "";
  }};
  
  &:hover{
    border: 3px solid black;
    border-radius: 5px;
  }
`;

const CheckButton = styled(Btn)`
  color: lightgreen;
  margin-right: 20px;
`;

const P = styled.p<StyledProps>`
  margin: 0;
  text-align: left;
  width: 100%;
  ${props => {
    if (props?.isComplete){
        return `
            text-decoration: line-through;
        `
    } else return "";
  }};
  
`;
//endregion

/**
 * Component which represents a single todo item in the list
 * @param item - the item of the todo list
 * @param removeFunc - function to remove item from the list
 * @constructor
 */
const TodoItem = ({item, removeFunc, toggleComplete}: TodoItemProps): JSX.Element => {
    return (
        <Container onClick={() => toggleComplete()} isComplete={item.isComplete}>
            <CheckButton style={item.isComplete ? {} : {opacity: 0}}><FaCheck/></CheckButton>
            <P isComplete={item.isComplete}>{item.text}</P>
            <Button onClick={(e) => removeFunc(item.id)}><FaTimes/></Button>
        </Container>
    );
};

export default TodoItem;