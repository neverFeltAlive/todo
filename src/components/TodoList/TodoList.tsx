import React, {useEffect, useState} from 'react';
import {
    FaCheck,
    FaSortAmountUpAlt,
    FaSortAmountDown,
    FaSortAlphaDown,
    FaSpellCheck,
    FaPlus,
    FaMinus
} from "react-icons/fa";
import TodoItem from "../TodoItem/TodoItem";
import styled from "styled-components";
import {Button as Btn} from "../UI";

//region Type
export type TodoItemType = {
    id: number
    text: string,
    isComplete: boolean,
}

type SortButtonProps = {
    color: string,
    isActive: boolean
}

enum Sorting {
    NONE,
    COMPLETE_FIRST,
    COMPLETE_LAST,
    ALPHA
}

enum Filter {
    NONE,
    COMPLETE,
    INCOMPLETE
}

type FilterType = (item: TodoItemType) => boolean;
//endregion

//region Style
const Input = styled.input`
  border: none;
  outline: none;
  padding: 5px;
  background-color: transparent;
  color: whitesmoke;
  font-size: 16px;
  font-weight: 700;

  @media (min-width: 1200px){
    font-size: 25px;
  }
`;

export const Button = styled(Btn)`
  color: lightblue;
`;

export const FooterButton = styled(Btn)<SortButtonProps>`
  color: ${props => props.color};
  opacity: ${props => props.isActive ? 1 : 0.3};
`;

const Fieldset = styled.fieldset`
  border: none;
`;

const InputFieldset = styled(Fieldset)`
  display: flex;
  justify-content: space-between;
  align-items: center;
  border-bottom: 2px solid gray;
  padding: 0;
  
  @media (min-width: 700px){
    min-width: 500px;
  }
  @media (min-width: 990px){
    min-width: 600px;
  }
  @media (min-width: 1200px){
    min-width: 800px;
  }
`;

const ListFieldset = styled(Fieldset)`
  padding: 10px 0 10px 0;
  margin-bottom: 80px;
  min-height: 200px;

  @media (min-width: 700px){
    min-height: 300px;
  }
  @media (min-width: 1200px){
    padding: 20px 0 20px 0;
    min-height: 500px;
  }
`;

const Footer = styled.div`
  display: flex;
  align-items: center;
  justify-content: space-between;
  position: absolute;
  bottom: 0;
  width: 100%;
`;

const FooterBlock = styled.div`
  display: flex;
  align-items: center;
`;

const FooterText = styled.p`
  margin: 0;
  color: gray;
`;
//endregion

//region Data
const filtersMap = new Map<Filter, FilterType>([
    [Filter.NONE, (item: TodoItemType) => true],
    [Filter.COMPLETE, (item: TodoItemType) => item.isComplete],
    [Filter.INCOMPLETE, (item: TodoItemType) => !item.isComplete],
]);

export const TEST_DATA = {
    input: "input",
    inputButton: "inputButton",
}
//endregion

/**
 * Component which represents a list of todo items and a form to add them
 * @constructor
 */
const TodoList = (): JSX.Element => {
    const [items, setItems] = useState<null | TodoItemType[]>(null);
    const [value, setValue] = useState<string>("");
    const [sorting, setSorting] = useState<Sorting>(Sorting.NONE)
    const [filters, setFilters] = useState<Filter[]>([]);

    /**
     * Modifies state by adding new todo text
     * @param text - string which needs to be added
     */
    const addItem = (text: string) => {
        if (text) {
            setItems(prevState => {
                const newItem = {id: Date.now(), text, isComplete: false};

                if (prevState) return [...prevState, newItem]
                else return [newItem]
            })
            setValue("");
        }
    }

    /**
     * Modifies state by removing the item
     * @param id - id of an item to be removed
     */
    const removeItem = (id: number) => {
        setItems((prevState) => {
            if (prevState) return prevState.filter(item => item.id !== id);
            else return null
        })
    }

    /**
     * Closure over a function which completes certain todo item
     * @param id - id of an item to be completed
     */
    const toggleCompleteItem = (id: number) => {
        return () => {
            setItems(prevState => {
                if (prevState && prevState.length > 0) {
                    const updateIndex = prevState.findIndex(item => item.id === id)
                    if (updateIndex !== undefined && updateIndex >= 0) {
                        const newItem = {...prevState[updateIndex], isComplete: !prevState[updateIndex].isComplete}
                        const newState = [...prevState];
                        newState[updateIndex] = newItem;
                        return newState
                    } else return prevState;
                } else return prevState;
            })
        }
    }

    /**
     * Toggles sorting mode
     * @param sort - sorting mode to toggle
     */
    const toggleSorting = (sort: Sorting) => {
        if (sorting === sort) setSorting(Sorting.NONE);
        else setSorting(sort);
    }

    /**
     * Applies sorting to the list of items
     * @param items - items to be sorted
     */
    const applySorting = (items: TodoItemType[]): TodoItemType[] => {
        switch (sorting) {
            case Sorting.NONE:
                return items;
            case Sorting.COMPLETE_LAST:
                return [...items].sort((a, b) => +a.isComplete - +b.isComplete)
            case Sorting.COMPLETE_FIRST:
                return [...items].sort((a, b) => -(+a.isComplete - +b.isComplete))
            case Sorting.ALPHA:
                return [...items].sort((a, b) => {
                    if (a.text !== b.text) {
                        if (a.text > b.text) return 1
                        else return -1;
                    } else return 0;
                })
            default:
                return items;
        }
    }

    /**
     * Toggles filtering mode
     * @param filt - filtering mode to toggle
     */
    const toggleFilter = (filt: Filter) => {
        if (!filters.includes(filt)) setFilters(prevState => [...prevState, filt]);
        else setFilters(prevState => {
            if (prevState) return prevState?.filter(item => item !== filt)
            else return prevState
        })
    }

    /**
     * Applies all filters to items list
     * @param items - items to be filtered
     */
    const applyFilters = (items: TodoItemType[]): TodoItemType[] => {
        let newList = items;
        for (const filt of filters) {
            const filterFunc = filtersMap.get(filt);
            if (filterFunc) newList = newList?.filter(filterFunc)
        }
        return newList;
    }

    // /**
    //  * Handles blur event on input element
    //  * @param event
    //  */
    // const handleBlur: React.FocusEventHandler<HTMLInputElement> = (event) =>
    //     addItem(event.target.value)
    //
    // /**
    //  * Handles Enter Press on input element
    //  * @param event
    //  */
    // const handleKeyDown: React.KeyboardEventHandler<HTMLInputElement> = (event) => {
    //     if (event.key === "Enter") {
    //         addItem(event.currentTarget.value)
    //         event.currentTarget.value = "";
    //     }
    // }

    return (
        <form onSubmit={(e) => e.preventDefault()} style={{position: "relative"}}>
            <InputFieldset>
                <Input type="text" value={value} onChange={(e) => setValue(e.target.value)}
                       placeholder="What needs to be done?" data-testid={TEST_DATA.input}/>
                <Button onClick={(e) => addItem(value)} data-testid={TEST_DATA.inputButton}><FaPlus/></Button>
            </InputFieldset>
            <ListFieldset>
                {items && applySorting(applyFilters(items))?.map((item, index) => {
                    return (
                        <TodoItem key={item.id} item={item} removeFunc={removeItem}
                                  toggleComplete={toggleCompleteItem(item.id)}/>
                    )
                })}
            </ListFieldset>
            <Footer>
                <FooterBlock>
                    <FooterText>Sort:</FooterText>
                    <FooterButton color="lightgreen" onClick={() => toggleSorting(Sorting.COMPLETE_FIRST)}
                                  isActive={sorting === Sorting.COMPLETE_FIRST}><FaSortAmountDown/></FooterButton>
                    <FooterButton color="lightcoral" onClick={() => toggleSorting(Sorting.COMPLETE_LAST)}
                                  isActive={sorting === Sorting.COMPLETE_LAST}><FaSortAmountUpAlt/></FooterButton>
                    <FooterButton color="lightblue" onClick={() => toggleSorting(Sorting.ALPHA)}
                                  isActive={sorting === Sorting.ALPHA}><FaSortAlphaDown/></FooterButton>
                </FooterBlock>
                <FooterBlock>
                    <FooterText>Filter:</FooterText>
                    <FooterButton color="lightblue" onClick={() => toggleFilter(Filter.COMPLETE)}
                                  isActive={filters.includes(Filter.COMPLETE)}><FaSpellCheck/></FooterButton>
                    <FooterButton color="lightcoral" onClick={() => toggleFilter(Filter.INCOMPLETE)}
                                  isActive={filters.includes(Filter.INCOMPLETE)}><FaMinus/></FooterButton>
                </FooterBlock>
            </Footer>
        </form>
    );
};

export default TodoList;