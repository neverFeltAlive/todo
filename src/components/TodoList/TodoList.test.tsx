import {render, screen} from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import TodoList, {TEST_DATA} from "./TodoList";
import {TEST_DATA as ITEM_TEST_DATA} from "../TodoItem/TodoItem";

describe("TodoList component", () => {
    test("TodoItem is added correctly", () => {
        render(<TodoList/>);
        const newItemText = "New item";

        userEvent.type(screen.getByTestId(TEST_DATA.input), newItemText);
        userEvent.click(screen.getByTestId(TEST_DATA.inputButton));
        expect(screen.getByText(newItemText)).toBeInTheDocument();
    });

    test("TodoItem is removed correctly", () => {
        render(<TodoList/>);
        const newItemText = "New item";

        userEvent.type(screen.getByTestId(TEST_DATA.input), newItemText);
        userEvent.click(screen.getByTestId(TEST_DATA.inputButton));
        expect(screen.getByText(newItemText)).toBeInTheDocument();

        userEvent.click(screen.getByTestId(ITEM_TEST_DATA.removeButton));
        expect(screen.queryByText(newItemText)).toBeNull();
    });

    test("TodoList matches snapshot", () => {
        const list = render(<TodoList/>);

        expect(list).toMatchSnapshot();
    });
});