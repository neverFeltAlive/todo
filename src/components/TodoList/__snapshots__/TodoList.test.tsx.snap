// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoList component TodoList matches snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <form
        style="position: relative;"
      >
        <fieldset
          class="sc-ftTHYK sc-pyfCe fUQwOd hYSmPh"
        >
          <input
            class="sc-jSUZER cvnciE"
            data-testid="input"
            placeholder="What needs to be done?"
            type="text"
            value=""
          />
          <button
            class="sc-bcXHqe sc-gKPRtg gUaoCe gHuEqz"
            data-testid="inputButton"
          >
            <svg
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 448 512"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M416 208H272V64c0-17.67-14.33-32-32-32h-32c-17.67 0-32 14.33-32 32v144H32c-17.67 0-32 14.33-32 32v32c0 17.67 14.33 32 32 32h144v144c0 17.67 14.33 32 32 32h32c17.67 0 32-14.33 32-32V304h144c17.67 0 32-14.33 32-32v-32c0-17.67-14.33-32-32-32z"
              />
            </svg>
          </button>
        </fieldset>
        <fieldset
          class="sc-ftTHYK sc-jrcTuL fUQwOd frDCqI"
        />
        <div
          class="sc-kDvujY ciFVxl"
        >
          <div
            class="sc-ipEyDJ dVJfLL"
          >
            <p
              class="sc-csuSiG SnRAC"
            >
              Sort:
            </p>
            <button
              class="sc-bcXHqe sc-iBYQkv gUaoCe emAOKu"
              color="lightgreen"
            >
              <svg
                fill="currentColor"
                height="1em"
                stroke="currentColor"
                stroke-width="0"
                viewBox="0 0 512 512"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M304 416h-64a16 16 0 0 0-16 16v32a16 16 0 0 0 16 16h64a16 16 0 0 0 16-16v-32a16 16 0 0 0-16-16zm-128-64h-48V48a16 16 0 0 0-16-16H80a16 16 0 0 0-16 16v304H16c-14.19 0-21.37 17.24-11.29 27.31l80 96a16 16 0 0 0 22.62 0l80-96C197.35 369.26 190.22 352 176 352zm256-192H240a16 16 0 0 0-16 16v32a16 16 0 0 0 16 16h192a16 16 0 0 0 16-16v-32a16 16 0 0 0-16-16zm-64 128H240a16 16 0 0 0-16 16v32a16 16 0 0 0 16 16h128a16 16 0 0 0 16-16v-32a16 16 0 0 0-16-16zM496 32H240a16 16 0 0 0-16 16v32a16 16 0 0 0 16 16h256a16 16 0 0 0 16-16V48a16 16 0 0 0-16-16z"
                />
              </svg>
            </button>
            <button
              class="sc-bcXHqe sc-iBYQkv gUaoCe hTeZji"
              color="lightcoral"
            >
              <svg
                fill="currentColor"
                height="1em"
                stroke="currentColor"
                stroke-width="0"
                viewBox="0 0 512 512"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M240 96h64a16 16 0 0 0 16-16V48a16 16 0 0 0-16-16h-64a16 16 0 0 0-16 16v32a16 16 0 0 0 16 16zm0 128h128a16 16 0 0 0 16-16v-32a16 16 0 0 0-16-16H240a16 16 0 0 0-16 16v32a16 16 0 0 0 16 16zm256 192H240a16 16 0 0 0-16 16v32a16 16 0 0 0 16 16h256a16 16 0 0 0 16-16v-32a16 16 0 0 0-16-16zm-256-64h192a16 16 0 0 0 16-16v-32a16 16 0 0 0-16-16H240a16 16 0 0 0-16 16v32a16 16 0 0 0 16 16zM16 160h48v304a16 16 0 0 0 16 16h32a16 16 0 0 0 16-16V160h48c14.21 0 21.39-17.24 11.31-27.31l-80-96a16 16 0 0 0-22.62 0l-80 96C-5.35 142.74 1.78 160 16 160z"
                />
              </svg>
            </button>
            <button
              class="sc-bcXHqe sc-iBYQkv gUaoCe hSlKhv"
              color="lightblue"
            >
              <svg
                fill="currentColor"
                height="1em"
                stroke="currentColor"
                stroke-width="0"
                viewBox="0 0 448 512"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M176 352h-48V48a16 16 0 0 0-16-16H80a16 16 0 0 0-16 16v304H16c-14.19 0-21.36 17.24-11.29 27.31l80 96a16 16 0 0 0 22.62 0l80-96C197.35 369.26 190.22 352 176 352zm240-64H288a16 16 0 0 0-16 16v32a16 16 0 0 0 16 16h56l-61.26 70.45A32 32 0 0 0 272 446.37V464a16 16 0 0 0 16 16h128a16 16 0 0 0 16-16v-32a16 16 0 0 0-16-16h-56l61.26-70.45A32 32 0 0 0 432 321.63V304a16 16 0 0 0-16-16zm31.06-85.38l-59.27-160A16 16 0 0 0 372.72 32h-41.44a16 16 0 0 0-15.07 10.62l-59.27 160A16 16 0 0 0 272 224h24.83a16 16 0 0 0 15.23-11.08l4.42-12.92h71l4.41 12.92A16 16 0 0 0 407.16 224H432a16 16 0 0 0 15.06-21.38zM335.61 144L352 96l16.39 48z"
                />
              </svg>
            </button>
          </div>
          <div
            class="sc-ipEyDJ dVJfLL"
          >
            <p
              class="sc-csuSiG SnRAC"
            >
              Filter:
            </p>
            <button
              class="sc-bcXHqe sc-iBYQkv gUaoCe hSlKhv"
              color="lightblue"
            >
              <svg
                fill="currentColor"
                height="1em"
                stroke="currentColor"
                stroke-width="0"
                viewBox="0 0 576 512"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M272 256h91.36c43.2 0 82-32.2 84.51-75.34a79.82 79.82 0 0 0-25.26-63.07 79.81 79.81 0 0 0 9.06-44.91C427.9 30.57 389.3 0 347 0h-75a16 16 0 0 0-16 16v224a16 16 0 0 0 16 16zm40-200h40a24 24 0 0 1 0 48h-40zm0 96h56a24 24 0 0 1 0 48h-56zM155.12 22.25A32 32 0 0 0 124.64 0H99.36a32 32 0 0 0-30.48 22.25L.59 235.73A16 16 0 0 0 16 256h24.93a16 16 0 0 0 15.42-11.73L68.29 208h87.42l11.94 36.27A16 16 0 0 0 183.07 256H208a16 16 0 0 0 15.42-20.27zM89.37 144L112 75.3l22.63 68.7zm482 132.48l-45.21-45.3a15.88 15.88 0 0 0-22.59 0l-151.5 151.5-55.41-55.5a15.88 15.88 0 0 0-22.59 0l-45.3 45.3a16 16 0 0 0 0 22.59l112 112.21a15.89 15.89 0 0 0 22.6 0l208-208.21a16 16 0 0 0-.02-22.59z"
                />
              </svg>
            </button>
            <button
              class="sc-bcXHqe sc-iBYQkv gUaoCe hTeZji"
              color="lightcoral"
            >
              <svg
                fill="currentColor"
                height="1em"
                stroke="currentColor"
                stroke-width="0"
                viewBox="0 0 448 512"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M416 208H32c-17.67 0-32 14.33-32 32v32c0 17.67 14.33 32 32 32h384c17.67 0 32-14.33 32-32v-32c0-17.67-14.33-32-32-32z"
                />
              </svg>
            </button>
          </div>
        </div>
      </form>
    </div>
  </body>,
  "container": <div>
    <form
      style="position: relative;"
    >
      <fieldset
        class="sc-ftTHYK sc-pyfCe fUQwOd hYSmPh"
      >
        <input
          class="sc-jSUZER cvnciE"
          data-testid="input"
          placeholder="What needs to be done?"
          type="text"
          value=""
        />
        <button
          class="sc-bcXHqe sc-gKPRtg gUaoCe gHuEqz"
          data-testid="inputButton"
        >
          <svg
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 448 512"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M416 208H272V64c0-17.67-14.33-32-32-32h-32c-17.67 0-32 14.33-32 32v144H32c-17.67 0-32 14.33-32 32v32c0 17.67 14.33 32 32 32h144v144c0 17.67 14.33 32 32 32h32c17.67 0 32-14.33 32-32V304h144c17.67 0 32-14.33 32-32v-32c0-17.67-14.33-32-32-32z"
            />
          </svg>
        </button>
      </fieldset>
      <fieldset
        class="sc-ftTHYK sc-jrcTuL fUQwOd frDCqI"
      />
      <div
        class="sc-kDvujY ciFVxl"
      >
        <div
          class="sc-ipEyDJ dVJfLL"
        >
          <p
            class="sc-csuSiG SnRAC"
          >
            Sort:
          </p>
          <button
            class="sc-bcXHqe sc-iBYQkv gUaoCe emAOKu"
            color="lightgreen"
          >
            <svg
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 512 512"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M304 416h-64a16 16 0 0 0-16 16v32a16 16 0 0 0 16 16h64a16 16 0 0 0 16-16v-32a16 16 0 0 0-16-16zm-128-64h-48V48a16 16 0 0 0-16-16H80a16 16 0 0 0-16 16v304H16c-14.19 0-21.37 17.24-11.29 27.31l80 96a16 16 0 0 0 22.62 0l80-96C197.35 369.26 190.22 352 176 352zm256-192H240a16 16 0 0 0-16 16v32a16 16 0 0 0 16 16h192a16 16 0 0 0 16-16v-32a16 16 0 0 0-16-16zm-64 128H240a16 16 0 0 0-16 16v32a16 16 0 0 0 16 16h128a16 16 0 0 0 16-16v-32a16 16 0 0 0-16-16zM496 32H240a16 16 0 0 0-16 16v32a16 16 0 0 0 16 16h256a16 16 0 0 0 16-16V48a16 16 0 0 0-16-16z"
              />
            </svg>
          </button>
          <button
            class="sc-bcXHqe sc-iBYQkv gUaoCe hTeZji"
            color="lightcoral"
          >
            <svg
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 512 512"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M240 96h64a16 16 0 0 0 16-16V48a16 16 0 0 0-16-16h-64a16 16 0 0 0-16 16v32a16 16 0 0 0 16 16zm0 128h128a16 16 0 0 0 16-16v-32a16 16 0 0 0-16-16H240a16 16 0 0 0-16 16v32a16 16 0 0 0 16 16zm256 192H240a16 16 0 0 0-16 16v32a16 16 0 0 0 16 16h256a16 16 0 0 0 16-16v-32a16 16 0 0 0-16-16zm-256-64h192a16 16 0 0 0 16-16v-32a16 16 0 0 0-16-16H240a16 16 0 0 0-16 16v32a16 16 0 0 0 16 16zM16 160h48v304a16 16 0 0 0 16 16h32a16 16 0 0 0 16-16V160h48c14.21 0 21.39-17.24 11.31-27.31l-80-96a16 16 0 0 0-22.62 0l-80 96C-5.35 142.74 1.78 160 16 160z"
              />
            </svg>
          </button>
          <button
            class="sc-bcXHqe sc-iBYQkv gUaoCe hSlKhv"
            color="lightblue"
          >
            <svg
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 448 512"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M176 352h-48V48a16 16 0 0 0-16-16H80a16 16 0 0 0-16 16v304H16c-14.19 0-21.36 17.24-11.29 27.31l80 96a16 16 0 0 0 22.62 0l80-96C197.35 369.26 190.22 352 176 352zm240-64H288a16 16 0 0 0-16 16v32a16 16 0 0 0 16 16h56l-61.26 70.45A32 32 0 0 0 272 446.37V464a16 16 0 0 0 16 16h128a16 16 0 0 0 16-16v-32a16 16 0 0 0-16-16h-56l61.26-70.45A32 32 0 0 0 432 321.63V304a16 16 0 0 0-16-16zm31.06-85.38l-59.27-160A16 16 0 0 0 372.72 32h-41.44a16 16 0 0 0-15.07 10.62l-59.27 160A16 16 0 0 0 272 224h24.83a16 16 0 0 0 15.23-11.08l4.42-12.92h71l4.41 12.92A16 16 0 0 0 407.16 224H432a16 16 0 0 0 15.06-21.38zM335.61 144L352 96l16.39 48z"
              />
            </svg>
          </button>
        </div>
        <div
          class="sc-ipEyDJ dVJfLL"
        >
          <p
            class="sc-csuSiG SnRAC"
          >
            Filter:
          </p>
          <button
            class="sc-bcXHqe sc-iBYQkv gUaoCe hSlKhv"
            color="lightblue"
          >
            <svg
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 576 512"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M272 256h91.36c43.2 0 82-32.2 84.51-75.34a79.82 79.82 0 0 0-25.26-63.07 79.81 79.81 0 0 0 9.06-44.91C427.9 30.57 389.3 0 347 0h-75a16 16 0 0 0-16 16v224a16 16 0 0 0 16 16zm40-200h40a24 24 0 0 1 0 48h-40zm0 96h56a24 24 0 0 1 0 48h-56zM155.12 22.25A32 32 0 0 0 124.64 0H99.36a32 32 0 0 0-30.48 22.25L.59 235.73A16 16 0 0 0 16 256h24.93a16 16 0 0 0 15.42-11.73L68.29 208h87.42l11.94 36.27A16 16 0 0 0 183.07 256H208a16 16 0 0 0 15.42-20.27zM89.37 144L112 75.3l22.63 68.7zm482 132.48l-45.21-45.3a15.88 15.88 0 0 0-22.59 0l-151.5 151.5-55.41-55.5a15.88 15.88 0 0 0-22.59 0l-45.3 45.3a16 16 0 0 0 0 22.59l112 112.21a15.89 15.89 0 0 0 22.6 0l208-208.21a16 16 0 0 0-.02-22.59z"
              />
            </svg>
          </button>
          <button
            class="sc-bcXHqe sc-iBYQkv gUaoCe hTeZji"
            color="lightcoral"
          >
            <svg
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 448 512"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M416 208H32c-17.67 0-32 14.33-32 32v32c0 17.67 14.33 32 32 32h384c17.67 0 32-14.33 32-32v-32c0-17.67-14.33-32-32-32z"
              />
            </svg>
          </button>
        </div>
      </div>
    </form>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
